src/util/worker/mod.rs

```diff
@@ -145,32 +157,36 @@ pub struct Worker<T: Display> {
     scheduler: Scheduler<T>,
     receiver: Mutex<Option<Receiver<Option<T>>>>,
     handle: Option<JoinHandle<()>>,
+    tasks_per_tick: Option<usize>,
 }
 
-fn poll<R, T>(mut runner: R, rx: Receiver<Option<T>>, counter: Arc<AtomicUsize>, batch_size: usize)
-where
+fn poll<R, T>(
+    mut runner: R,
+    rx: Receiver<Option<T>>,
+    counter: Arc<AtomicUsize>,
+    batch_size: usize,
+    tasks_per_tick: Option<usize>,
+) where
     R: BatchRunnable<T> + Send + 'static,
     T: Display + Send + 'static,
 {
     let name = thread::current().name().unwrap().to_owned();
+    let tasks_per_tick = tasks_per_tick.unwrap_or(0);
+    let mut timing_task_counter = 0;
     let mut keep_going = true;
     let mut buffer = Vec::with_capacity(batch_size);
     while keep_going {
-        let t = rx.recv();
-        match t {
-            Ok(Some(t)) => buffer.push(t),
-            _ => break,
-        }
-        while buffer.len() < batch_size {
-            match rx.try_recv() {
-                Ok(None) => {
-                    keep_going = false;
-                    break;
-                }
-                Ok(Some(t)) => buffer.push(t),
-                _ => break,
-            }
+        keep_going = next_batch(
+            &rx,
+            &mut runner,
+            &mut buffer,
+            batch_size,
+            &mut timing_task_counter,
```

Comment:
you don't need to pass timing_task_counter, you can increase it by the buffer length. 

Btw, tasks_per_tick should be for the tasks number we received.